File: SuperheroApp/DataSource/Remote/CharactersRemoteDataSource.m
//
//  CharactersRemoteDataSource.m
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 04..
//  Copyright © 2018. W.UP. All rights reserved.
//

#import "CharactersRemoteDataSource.h"
#import "NSString+MD5.h"
#import "SuperheroApp-Swift.h"
#import <AFNetworking/AFNetworking.h>
#import <CommonCrypto/CommonDigest.h>

#define PUBLIC_API_KEY      [[NSBundle mainBundle] objectForInfoDictionaryKey:@"MarvelPublicApiKey"]
#define PRIVATE_API_KEY     [[NSBundle mainBundle] objectForInfoDictionaryKey:@"MarvelPrivateApiKey"]

@interface CharactersRemoteDataSource()

@property(nonatomic, strong, readonly) AFHTTPSessionManager *sessionManager;

@end

@implementation CharactersRemoteDataSource

- (instancetype)init {
    AFHTTPSessionManager *sessionManager = [[AFHTTPSessionManager alloc] initWithBaseURL:[NSURL URLWithString:@"https://gateway.marvel.com/"]];
    
    sessionManager.requestSerializer = [AFHTTPRequestSerializer serializer];
    sessionManager.responseSerializer = [AFJSONResponseSerializer serializer];
    
    return [self initWithSessionManager:sessionManager];
}

- (instancetype)initWithSessionManager:(AFHTTPSessionManager *)sessionManager {
    self = [super init];
    
    if (self) {
        _sessionManager = sessionManager;
    }
    
    return self;
}

- (void)loadCharacter:(NSString *)characterId complete:(void (^)(Character *))complete error:(void (^)(void))error {

    NSString *url = [NSString stringWithFormat:@"v1/public/characters/%@", characterId];
    
    NSUInteger timestamp = [[NSDate new] timeIntervalSince1970];
    NSString *hash = [NSString stringWithFormat:@"%lu%@%@", (unsigned long)timestamp, PRIVATE_API_KEY, PUBLIC_API_KEY];
    
    NSDictionary *parameters = @{
                                 @"apikey": PUBLIC_API_KEY,
                                 @"ts": @(timestamp),
                                 @"hash": [hash md5]
                                 };
    
    [self.sessionManager GET: url parameters:parameters progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id _Nullable responseObject) {
        
        NSLog(@"response %@", responseObject);
        
        NSDictionary *data = [responseObject valueForKey:@"data"];
        NSArray *results = [data valueForKey:@"results"];
        Character *character = nil;
        
        NSDictionary *item = [results firstObject];
        
        if (item != nil) {
            character = [self extractCharacterFromJsonDisctionary:item];
        }
        
        if (complete) {
            complete(character);
        }
        
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull err) {
        if (error) {
            error();
        }
    }];
}

- (void)loadCharacters:(Page *)page complete:(void (^)(NSArray<Character *> *))complete error:(void (^)(void))error {
 
    NSUInteger timestamp = [[NSDate new] timeIntervalSince1970];
    NSString *hash = [NSString stringWithFormat:@"%lu%@%@", (unsigned long)timestamp, PRIVATE_API_KEY, PUBLIC_API_KEY];
    
    NSDictionary *parameters = @{
                                 @"limit": @(page.limit),
                                 @"offset": @(page.offset),
                                 @"apikey": PUBLIC_API_KEY,
                                 @"ts": @(timestamp),
                                 @"hash": [hash md5]
                                 };
    
    [self.sessionManager GET:@"v1/public/characters" parameters:parameters progress:nil success:^(NSURLSessionDataTask * _Nonnull task, id  _Nullable responseObject) {
        
        NSLog(@"response %@", responseObject);
        
        NSDictionary *data = [responseObject valueForKey:@"data"];
        NSArray *results = [data valueForKey:@"results"];
        NSMutableArray<Character *> *characters = [NSMutableArray arrayWithCapacity:results.count];
        
        for (NSDictionary *item in results) {
            
            Character *character = [self extractCharacterFromJsonDisctionary:item];
    
            [characters addObject:character];
        }
        
        if (complete) {
            complete(characters);
        }
        
    } failure:^(NSURLSessionDataTask * _Nullable task, NSError * _Nonnull err) {
        NSLog(@"Error during load characters: %@", [err localizedDescription]);
        NSData *response = err.userInfo[AFNetworkingOperationFailingURLResponseDataErrorKey];
        NSString *responseString = [[NSString alloc] initWithData:response encoding:NSUTF8StringEncoding];
        NSLog(@"Error response: %@", responseString);
        error();
    }];
}

- (void)saveCharacters:(NSArray<Character *> *)characters complete:(void (^)(void))complete error:(void (^)(void))error {
    complete();
}

- (Character *)extractCharacterFromJsonDisctionary:(NSDictionary *)item {
    Character *character = [Character new];
    character.name = [item valueForKey:@"name"];
    character.characterId = [item valueForKey:@"id"];
    
    NSDictionary *thumbnail = [item valueForKey:@"thumbnail"];
    NSString *thumbnailUrl = [NSString stringWithFormat:@"%@.%@", [thumbnail valueForKey:@"path"], [thumbnail valueForKey:@"extension"]];
    character.thumbnailUrl = [thumbnailUrl stringByReplacingOccurrencesOfString:@"http" withString:@"https"];
    
    return character;
}

@end

File: SuperheroApp/DataSource/Remote/CharactersRemoteDataSource.swift
//
//  CharactersRemoteDataSource.swift
//  SuperheroApp
//
//  Created by Balazs Varga on 2018. 02. 11..
//  Copyright © 2018. W.UP. All rights reserved.
//

import Alamofire
import CodableAlamofire
import AlamofireActivityLogger

class CharactersRemoteDataSource : CharactersDataSource {

    private let manager: SessionManager
    private let publicApiKey: String
    private let privateApiKey: String
    
    init(manager: SessionManager, publicApiKey: String, privateApiKey: String) {
        self.manager = manager
        self.publicApiKey = publicApiKey
        self.privateApiKey = privateApiKey
    }
    
    func loadCharacters(page: Page, complete: @escaping ([Character]) -> Void, fail: @escaping () -> Void) {
        
        let timestamp = Int(Date().timeIntervalSince1970)
        let hash = "\(timestamp)\(privateApiKey)\(publicApiKey)"
        
        let parameters = [ "limit": page.limit,
                           "offset": page.offset,
                           "apikey": self.publicApiKey,
                           "ts": timestamp,
                           "hash": hash.md5() ] as [String : Any]
        
    
        self.manager.request("https://gateway.marvel.com/v1/public/characters", parameters: parameters, encoding: URLEncoding.default)
            .validate()
            .log()
            .responseDecodableObject(keyPath: "data.results", decoder: JSONDecoder(), completionHandler: { (response: DataResponse<[Character]>) in
                
                if (response.error != nil) {
                    fail()
                } else {
                    complete(response.value!)
                }
            })
    }
    
    func loadCharacter(characterId: String, complete: @escaping (Character?) -> Void, fail: @escaping () -> Void) {
        let timestamp = Int(Date().timeIntervalSince1970)
        let hash = "\(timestamp)\(privateApiKey)\(publicApiKey)"
        
        let parameters = [ "apikey": self.publicApiKey,
                           "ts": timestamp,
                           "hash": hash.md5() ] as [String : Any]
        
        
        self.manager.request("https://gateway.marvel.com/v1/public/characters/\(characterId)", parameters: parameters, encoding: URLEncoding.default)
            .validate()
            .log()
            .responseDecodableObject(keyPath: "data.results", decoder: JSONDecoder(), completionHandler: { (response: DataResponse<[Character]>) in
                if (response.error != nil) {
                    fail()
                } else {
                    complete(response.value?.first)
                }
            })
    }
    
    func saveCharacters(characters: Array<Character>, complete: @escaping () -> Void, fail: @escaping () -> Void) {
        complete()
    }
}
